xds/internal/balancer/clusterresolver/resource_resolver.go

```diff
@@ -0,0 +1,236 @@
+/*
+ *
+ * Copyright 2021 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package clusterresolver
+
+import (
+	"reflect"
+	"sync"
+
+	"google.golang.org/grpc/xds/internal/balancer/clusterresolver/balancerconfig"
+	"google.golang.org/grpc/xds/internal/xdsclient"
+)
+
+// resourceUpdate is a combined updates from all the resources, in the order of
+// priority. For example, it can be {EDS, EDS, DNS}.
+type resourceUpdate struct {
+	p   []balancerconfig.PriorityConfig
+	err error
+}
+
+type resolverInterface interface {
+	lastUpdate() (interface{}, bool)
+	resolveNow()
+	stop()
+}
+
+// discoveryMechanismKey is {type+resource_name}, it's used as the map key, so
+// that the same resource resolver can be reused (e.g. when there are two
+// mechanisms, both for the same EDS resource, but has different circuit
+// breaking config.
+type discoveryMechanismKey struct {
+	typ  balancerconfig.DiscoveryMechanismType
+	name string
+}
+
+// resolverMechanismTuple is needed to keep the resolver and the discovery
+// mechanism together, because resolvers can be shared. And we need the
+// mechanism for fields like circuit breaking, LRS etc when generating the
+// balancer config.
+type resolverMechanismTuple struct {
+	dm    balancerconfig.DiscoveryMechanism
+	dmKey discoveryMechanismKey
+	r     resolverInterface
+}
+
+type resourceResolver struct {
+	parent        *clusterResolverBalancer
+	updateChannel chan *resourceUpdate
```

Comment:
Given that we seem to care only about the most recent update, what do you think about replacing this with a field and a mutex.

```golang
type resourceResolver struct {
  ...
  updateMu sync.Mutex
  latestUpdate *resourceUpdate
  ...
}

func (rr *resourceResolver) setUpdate(u *resourceUpdate) {
  // grab lock, update field, and release lock
}

func (rr *resourceResolver) getUpdate() *resourceUpdate {
}
```

This would have two advantages:
- the discovery mechanism instances will no longer have to reach and grab the lock on the `resourceResolver` as they do now
- we can skip the code which drains the channel before pushing into it from multiple places.

Reply:
Also, I see no one reading from this channel and acting on the update pushed to it. Will that come later on?